### ISSUE TITLE
Implement seekable streams

### ISSUE THREAD

**Issue by cdmurph32** [WebAssembly/component-model#418]
Seekable streams would enable a great deal of functionality for my [C2PA SDK](https://github.com/contentauth/c2pa-rs/tree/main/sdk) work and likely many other use cases. I don't have much of an understanding of how this would be implemented, but I'd like to start a conversation. 

The most obvious use case would be file handles. This would allow feature parity with [web_sys::FileSystemFileHandle](https://rustwasm.github.io/wasm-bindgen/api/web_sys/struct.FileSystemFileHandle.html).
For example this [JS code](https://github.com/contentauth/c2pa-bindings/blob/e3d76afc491f61c5a362599a5dae2b9a96d67955/libs/c2pa-js/src/lib/classes/Builder.ts#L53) uses this [wasm-bindgen rust code](https://github.com/contentauth/c2pa-bindings/blob/e3d76afc491f61c5a362599a5dae2b9a96d67955/libs/c2pa_wasm/src/wasm_builder.rs#L78-L85): 
```
 async signFileSystemHandle(
    signer: AsyncSigner,
    format: string,
    source: FileSystemFileHandle,
    dest: FileSystemFileHandle,
  ): Promise<Uint8Array> {
    const {
      value: { signerPtr, manifestBytes },
      ptr,
    } = await Builder.pool.builder_signFileSystemHandle(
      Builder.workerConfig,
      this.getPtr(),
      signer.getPtr(),
      format,
      source,
      dest,
    );
    this.ptr = ptr;
    signer.setPtr(signerPtr);

    return manifestBytes;
  }
```
The wasm-bindgen rust code can use FIleSystemFileHandle as a stream using [createWritable](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemFileHandle/createWritable)
```
    #[wasm_bindgen(js_name = signFileSystemHandle)]
    pub async fn sign_file_system_handle(
        mut self,
        signer_ptr: u32,
        format: &str,
        source_handle: &FileSystemFileHandle,
        dest_handle: &FileSystemFileHandle,
    ) -> Result<JsValue, JsSysError> {
        let signer = JavaScriptAsyncSigner::from_ptr(signer_ptr)?;
        let mut source = HandleStream::new(source_handle).await?;
        let mut dest = HandleStream::new(dest_handle).await?;
        let manifest_bytes = self
            .builder
            .sign_async(&signer, format, &mut source, &mut dest)
            .await
            .unwrap();
```

Besides file handles, seekable streams would be very useful for other use cases. C2PA manifests must be located in very large asset containers, which must be parsed to understand their structure. This requires the streams to be seekable. As a workaround, [my POC of a C2PA component](https://github.com/cdmurph32/c2pa-component/blob/v24/component/src/lib.rs#L205-L209) reads the input stream into a cursor before passing it to the c2pa SDK functions. This is obviously far from ideal and will not work with very large containers streams such as video.
```
fn add_seek_to_read<R: Read>(mut reader: R) -> std::io::Result<Cursor<Vec<u8>>> {
    let mut buffer = Vec::new();
    reader.read_to_end(&mut buffer)?;
    Ok(Cursor::new(buffer))
}
```

**Comment by dicej** [comment]
@lukewagner have you thought at all about how seeking would work for `stream<T>`?

**Comment by lukewagner** [comment]
In general, I've been assuming that seeking happens via methods on resources that return new streams, e.g.:
```wit
resource file {
  pread-stream: func(offset: u64) -> stream<u8>;
}
```
which reflects the fact that not all stream sources are seekable.

However, one operation that *does* make sense on a stream (and almost made it into wasi-io/streams, iirc) which I think makes sense as a canon built-in would be a `stream.skip` that in effect lets you "seek" forward N elements as an optimization over simply reading and discarding N elements.  I think this make sense for all streams.

**Comment by sunfishcode** [comment]
(For what it's worth, wasi-io/streams [does have `skip`](https://github.com/WebAssembly/wasi-io/blob/main/wit/streams.wit#L81) for seeking forward.)

**Comment by lukewagner** [comment]
Oops, right, I just missed it in my scan :P

**Comment by cdmurph32** [comment]
For wasi-io/streams `skip` and not `seek` makes a lot of sense, but for wasi::filesystem specifically it would be very useful to have seek. It would allow us to use wasi for our NodeJS SDK. Does that sound like a reasonable feature?

**Comment by dicej** [comment]
> For wasi-io/streams `skip` and not `seek` makes a lot of sense, but for wasi::filesystem specifically it would be very useful to have seek. It would allow us to use wasi for our NodeJS SDK. Does that sound like a reasonable feature?

Would the `pread-stream` sketch Luke posted above be suitable?  The idea is that you can't seek an existing stream, but you can open a new one at an arbitrary offset within the file.  And you can always build an abstraction on top of that which makes it look like you're seeking with a single stream.

**Comment by cdmurph32** [comment]
Yes. I think that sounds good. I wasn't aware that we already had what I was basically looking for with [descriptors](https://github.com/WebAssembly/wasi-filesystem/blob/0f2a9a95730d4d5b7b32b7f9777a230fc7893c80/wit/types.wit#L298).
Should we leave this open until filesystem 0.3.0, or just close it?

**Comment by sunfishcode** [comment]
I think we can close this, as wasi-filesystem 0.2.0 already uses this pattern of returning new streams at offsets rather than having a seek on `input-stream` or `output-stream` , so it's now what [wasi-filesystem 0.3.0](https://github.com/WebAssembly/wasi-filesystem/pull/164) is proposed to do as well, so I don't think there's anything else needed from the component-model repo side.
